text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,551 @@
+
+If the private key is protected by a password, the user will be prompted
+for a password interactively. Alternatively, the user could provide the
+password in an environment variable ``EDGEDB_PRIVATE_KEY_PASSWORD``. In
+either way, the password will be stored in the same ``metadata.json``
+and be further used to load the private key.
+
+Here is a proposed sample ``metadata.json``::
+
+    {
+        "format": 2,
+        "version": "1-beta2",
+        "current_version": "1.0b2+ga7130d5c7.cv202104290000",
+        "slot": "1-beta2",
+        "method": "Package",
+        "port": 10733,
+        "start_conf": "Auto",
+        "tls_certfile": "path/to/cert.pem",
+        "tls_keyfile": "path/to/key.pem",
+        "tls_private_key_password": "password-in-cleartext"
+    }
+
+The EdgeDB CLI will correspondingly generate a system service file that
+eventually launches the EdgeDB server as follows::
+
+    EDGEDB_TLS_PRIVATE_KEY_PASSWORD=password-in-cleartext \
+    edgedb-server \
+        --port=10733 \
+        --tls-certfile=path/to/cert.pem \
+        --tls-keyfile=path/to/key.pem
+
+This RFC does not involve setting up a proper CA-based trust chain for
+production usage. The knowledge will be well-documented, and products
+like the Aether will have its own way doing so.
+
+
+Generate Self-signed Certificates
+---------------------------------
+
+The EdgeDB CLI will automatically generate a self-signed certificate if
+``--tls-certfile`` is not present in command ``edgedb project init`` or
+``edgedb server init`` under the data directory of that EdgeDB instance
+as described in the previous section, in the names of
+``edgedb-cert.pem`` for the certificate in PEM format, and
+``edgedb-key.pem`` for the private key (no passphrase for simplicity).
+This certificate is supposed to be used for development purposes only.
+
+Likewise, the ``metadata.json`` file will be update with the full path
+to the generated files for consistency.
+
+
+Client-side Server Certificate Verification
+===========================================
+
+On the client side (both the language bindings and the REPL), TLS server
+certificate verification should always be enabled. In order to accept
+the self-signed certificate, at the time of certificate generation, the
+EdgeDB CLI will also copy the generated certificate into the so-called
+``credentials.json`` - a group of JSON files named after the EdgeDB
+instance in a well-known place (e.g. ``~/.edgedb/credentials/``
+depending on the OS) that are meant to store credentials for the client
+to establish connections to the EdgeDB instance. For example::
+
+    {
+        "port": 10732,
+        "user": "edgedb",
+        "password": "login-password-in-clear-text",
+        "database": "edgedb",
+        "tls_certdata": "-----BEGIN CERTIFICATE-----\nMIICvjCCAaagAw..."
+    }
+
+The language bindings and the REPL should load the certificate from the
+value of ``tls_certdata`` and trust only that certificate for connecting
```

Comment:
I'd clarify that `tls_certdata` can specify an issuing certificate as well up to some long-lived root cert.

Reply:
👍 Thanks!  Added an open question - looking forward to more comments on this topic.